agent/src/agentclient.rs

```diff
@@ -0,0 +1,173 @@
+use k8s_openapi::api::core::v1::Node;
+use kube::Api;
+use snafu::{futures::TryFutureExt, ResultExt};
+use std::env;
+use tokio::time::{sleep, Duration};
+
+use crate::apiclient::{boot_update, get_os_info, list_available, prepare, update};
+use crate::error::{self, Result};
+use apiserver::api::UpdateBottlerocketNodeRequest;
+use models::constants::NAMESPACE;
+use models::node::{
+    BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState, BottlerocketNodeStatus,
+};
+
+const AGENT_SLEEP_DURATION: Duration = Duration::from_millis(5000);
+
+#[derive(Clone)]
+pub struct BrupopAgent {
+    k8s_client: kube::client::Client,
+}
+
+impl BrupopAgent {
+    pub fn new(k8s_client: kube::client::Client) -> Self {
+        BrupopAgent { k8s_client }
+    }
+
+    pub async fn ensure_node_custom_resource_exists(&self) {
+        // Todo (tianhg)
+    }
+
+    async fn fetch_custom_resource(&self) -> Result<BottlerocketNode> {
+        // TODO Nit: This comment doesn't say much more than the line of code following it.
+        // It might be good to add more context about what the node name is, or what it's used for.
+        let associated_bottlerocketnode_name =
+            format!("brn-{}", self.get_node_selector().await?.node_name);
+
+        let bottlerocket_nodes: Api<BottlerocketNode> =
+            Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+        let associated_bottlerocketnode: BottlerocketNode = bottlerocket_nodes
+            .get(&associated_bottlerocketnode_name)
+            .context(error::FetchCustomResource)
+            .await?;
+
+        Ok(associated_bottlerocketnode)
+    }
+
+    async fn get_node_selector(&self) -> Result<BottlerocketNodeSelector> {
+        let associated_node_name = env::var("MY_NODE_NAME").context(error::GetNodeName)?;
```

Comment:
Is this going to be the actual environment variable name that customers will set up? Or is that env variable coming from the kubernetes cluster in a weird way?

Reply:
Yeah. It's a actual env variable name that set up by ourselves on `burpop-agent.yaml` file
```
env:
        - name: MY_NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
```